packages/legend-query-builder/src/stores/QueryLoaderState.ts

```diff
@@ -0,0 +1,357 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  DEFAULT_TYPEAHEAD_SEARCH_MINIMUM_SEARCH_LENGTH,
+  DEFAULT_TYPEAHEAD_SEARCH_LIMIT,
+  type GenericLegendApplicationStore,
+} from '@finos/legend-application';
+import {
+  type LightQuery,
+  QuerySearchSpecification,
+  type AbstractPureGraphManager,
+  type QueryInfo,
+  type QueryProjectCoordinates,
+  V1_PureGraphManager,
+  type Query,
+} from '@finos/legend-graph';
+import {
+  ActionState,
+  type GeneratorFn,
+  assertErrorThrown,
+  guaranteeType,
+  deleteEntry,
+  SerializationFactory,
+  usingModelSchema,
+  type PlainObject,
+  guaranteeNonNullable,
+} from '@finos/legend-shared';
+import { makeObservable, observable, action, flow, flowResult } from 'mobx';
+import {
+  QUERY_BUILDER_CACHE_KEY,
+  CACHE_RECENTLY_VIEWED_QUERIES_LIMIT,
+} from '../application/QueryBuilderCache.js';
+import { createModelSchema, list, optional, primitive } from 'serializr';
+import type { QueryBuilderState } from './QueryBuilderState.js';
+import type {
+  LoadQueryFilterOption,
+  QueryBuilder_LegendApplicationPlugin_Extension,
+} from './QueryBuilder_LegendApplicationPlugin_Extension.js';
+
+class QueryPersistInfo {
+  id!: string;
+  owner?: string | undefined;
+  static readonly serialization = new SerializationFactory(
+    createModelSchema(QueryPersistInfo, {
+      id: primitive(),
+      owner: optional(primitive()),
+    }),
+  );
+}
+
+class QueryStorageState {
+  persistedQueries: QueryPersistInfo[] = [];
+
+  static readonly serialization = new SerializationFactory(
+    createModelSchema(QueryStorageState, {
+      persistedQueries: list(
+        usingModelSchema(QueryPersistInfo.serialization.schema),
+      ),
+    }),
+  );
+
+  static create(json: PlainObject<QueryStorageState>): QueryStorageState {
+    return QueryStorageState.serialization.fromJson(json);
+  }
+
+  /**
+   * This function persists the queryIds in the local storage.
+   * Number of queryIds we want to persist is `CACHE_RECENTLY_VIEWED_QUERIES_LIMIT`.
+   * We always want the order in which the queries are stored in a way that
+   * queries owned and recently edited by users comes first, followed by
+   * recently viewed and owned by user, followed by queries recently viewed and not
+   * owned by the user.
+   */
+  persistQueryIds(
+    val: LightQuery | Query,
+    applicationStore: GenericLegendApplicationStore,
+    currentUserId: string | undefined,
+  ): void {
+    let persistQuery = false;
+    if (
```

Comment:
You know we should create a `Queue` data structure in `@finos/legend-shared` which supports `swap` operation and it will do the job here